### ISSUE TITLE
#5115 has changed predictions

### ISSUE THREAD

**Issue by Kodiologist** [dmlc/xgboost#5203]
I would have thought that under #5115, DART predictions would be the same as they had been before when `ntreelimit` was set to a nonzero number. However, it appears that DART now produces new predictions, which are unaffected by whether `ntreelimit` is set in `predict`, as desired, but are different from pre-#5115 predictions, both with `ntreelimit` and without. Is this a regression, or are the new predictions correct?

<details>
<summary>Example</summary>

This code is based on the test "SHAPs sum to predictions, with or without DART" in `R-package/tests/testthat/test_helpers.R`.

```
library(xgboost, lib.loc = "/home/hippo/Workspace/xgboost/mylib")

set.seed(15)

d <- cbind(
  x1 = rnorm(100),
  x2 = rnorm(100),
  x3 = rnorm(100))
y <- d[,"x1"] + d[,"x2"]^2 +
  ifelse(d[,"x3"] > .5, d[,"x3"]^2, 2^d[,"x3"]) +
  rnorm(100)
nrounds <- 30

fit <- xgboost(
  verbose = 0,
  params =
    list(
      booster = "dart",
      objective = "reg:linear",
      eval_metric = "rmse",
      rate_drop = .01,
      one_drop = T),
  data = d,
  label = y,
  nrounds = nrounds)

pr <- function(...)
  predict(fit, newdata = d, ...)
cat("w/o ntreelimit\n")
set.seed(2)
print(pr())
cat("with ntreelimit\n")
set.seed(3)
print(pr(ntreelimit = nrounds))
```

On 7b656981, this prints:

```
w/o ntreelimit
  [1]  0.92676473  2.23212862  0.92792749  4.42379999  2.46226597  0.69863331
  [7]  1.10363221  1.67683148  6.10005331 -0.05690449  6.36369705  0.98988873
 [13]  1.03287458 -0.91235888  3.15261078  2.70877647  3.19321156  1.32619429
 [19]  1.17865765  1.67584610  3.14247656  3.12332392  0.37124518  0.67973423
 [25]  4.51128864  0.52064961  1.49891758  1.33003640  0.27817956  4.45876408
 [31]  7.97304583  0.19485372  4.85256052  1.98937798  6.95747662  1.05429363
 [37]  0.81055045  1.54438281  3.40924335  1.51254141  1.83467782  0.92281181
 [43]  8.75543785  6.77227545  0.08744153  3.77030659  2.66593552  3.12332392
 [49]  5.88673639  1.80606377  3.55537844  1.41737854  1.57731307  3.85868001
 [55]  5.76396847  2.00494838  5.50364828  7.21862841  0.65110290  0.16672906
 [61]  0.36554265  0.25567573  2.44472027 -1.00277030 -0.37950361 -1.13353014
 [67]  0.20976081  8.21963120  1.82166123  2.87193871 -1.13127887 -0.70636380
 [73]  1.43734384  2.99790430  0.32195681  2.38616228  6.44007301  1.20697117
 [79]  0.97582287  4.70257759  0.69335413  4.50990772  2.28238773  1.93220294
 [85]  1.06902552  2.35184860  1.99651718  1.52437580  0.35713023  1.20308709
 [91]  0.24007803  3.16458821  3.30302906  5.99578714  1.98643124  4.61541176
 [97]  1.07452476  2.51957130  0.11720890 -0.66286957
with ntreelimit
  [1]  0.93783569  2.24319935  0.93899852  4.43487072  2.47333717  0.68969464
  [7]  1.18220854  1.66789281  6.11112404 -0.04583353  6.45779181  1.00095963
 [13]  1.04394555 -0.92129755  3.16368175  2.71984744  3.20428252  1.33726525
 [19]  1.16971898  1.68691707  3.15354753  3.13439488  0.36230648  0.69080514
 [25]  4.50234985  0.46681473  1.48997903  1.34110737  0.26924086  4.46983528
 [31]  7.98411655  0.18591502  4.84362173  1.98043931  7.03781223  1.04535496
 [37]  0.82162142  1.53544414  3.42031431  1.59111750  1.84574878  0.91394043
 [43]  8.88343811  6.76333666  0.07850283  3.78137755  2.65699673  3.13439488
 [49]  5.89780712  1.81713474  3.56644940  1.42844951  1.58838403  3.86975098
 [55]  5.77503920  2.01601934  5.51471901  7.22969913  0.66217387  0.15779036
 [61]  0.37661365  0.24673703  2.43578172 -1.01170897 -0.38844228 -1.12245917
 [67]  0.20082211  8.23070240  1.83273220  2.88300967 -1.12020791 -0.71530247
 [73]  1.42840517  3.00897527  0.26812190  2.46473837  6.45114374  1.19803238
 [79]  0.96688414  4.71364880  0.70442510  4.50096893  2.29345846  1.94327390
 [85]  1.08009648  2.34291005  2.00758815  1.51543725  0.34784365  1.28166342
 [91]  0.23113936  3.17565918  3.31410003  6.00444269  1.99750221  4.62648296
 [97]  1.08559573  2.51063275  0.10827023 -0.65179861
```

On master, this prints

```
w/o ntreelimit
  [1]  0.8422105  2.2793722  0.8825806  4.4846115  2.4932570  0.7701296
  [7]  1.1263112  1.5966307  6.1777277 -0.1132701  6.3833289  1.0663240
 [13]  1.0500160 -1.0413350  3.1556480  2.7304411  3.2307172  1.5820167
 [19]  1.1922027  1.5404651  3.0971682  3.1875584  0.3140902  0.6787448
 [25]  4.5342412  0.5487907  1.6463656  1.3223281  0.2940366  4.4831362
 [31]  7.9416385  0.1396081  4.8498306  1.9636259  6.9444747  1.0179384
 [37]  0.7881684  1.5060399  3.4849648  1.5957783  1.7017839  1.0439169
 [43]  8.8352699  6.8884521  0.2274487  3.6971297  2.5306225  3.2063131
 [49]  5.9057541  1.8102639  3.5413160  1.6122520  1.6925395  3.8383603
 [55]  5.7502294  2.0324926  5.4687400  7.2141447  0.4488493  0.1699183
 [61]  0.5751100  0.2719634  2.4139721 -0.9163762 -0.3023523 -1.0377461
 [67]  0.1932511  8.1510162  1.8593879  2.8779652 -0.9259645 -0.7158566
 [73]  1.4541243  2.9631851  0.2720932  2.3893008  6.4605975  1.1576927
 [79]  0.9339637  4.7624793  0.7413906  4.5066066  2.3323772  1.9024912
 [85]  1.1373091  2.3808026  1.9702703  1.3333858  0.3582118  1.2185564
 [91]  0.2182850  3.0462012  3.2532260  5.9983945  2.0317540  4.6412554
 [97]  0.8871353  2.5109260  0.1824817 -0.7095767
with ntreelimit
  [1]  0.8422105  2.2793722  0.8825806  4.4846115  2.4932570  0.7701296
  [7]  1.1263112  1.5966307  6.1777277 -0.1132701  6.3833289  1.0663240
 [13]  1.0500160 -1.0413350  3.1556480  2.7304411  3.2307172  1.5820167
 [19]  1.1922027  1.5404651  3.0971682  3.1875584  0.3140902  0.6787448
 [25]  4.5342412  0.5487907  1.6463656  1.3223281  0.2940366  4.4831362
 [31]  7.9416385  0.1396081  4.8498306  1.9636259  6.9444747  1.0179384
 [37]  0.7881684  1.5060399  3.4849648  1.5957783  1.7017839  1.0439169
 [43]  8.8352699  6.8884521  0.2274487  3.6971297  2.5306225  3.2063131
 [49]  5.9057541  1.8102639  3.5413160  1.6122520  1.6925395  3.8383603
 [55]  5.7502294  2.0324926  5.4687400  7.2141447  0.4488493  0.1699183
 [61]  0.5751100  0.2719634  2.4139721 -0.9163762 -0.3023523 -1.0377461
 [67]  0.1932511  8.1510162  1.8593879  2.8779652 -0.9259645 -0.7158566
 [73]  1.4541243  2.9631851  0.2720932  2.3893008  6.4605975  1.1576927
 [79]  0.9339637  4.7624793  0.7413906  4.5066066  2.3323772  1.9024912
 [85]  1.1373091  2.3808026  1.9702703  1.3333858  0.3582118  1.2185564
 [91]  0.2182850  3.0462012  3.2532260  5.9983945  2.0317540  4.6412554
 [97]  0.8871353  2.5109260  0.1824817 -0.7095767
```

</details>

**Comment by trivialfis** [comment]
Yup. Before the prediction relies on `ntree_limit` to distinguish whether should we perform dropout, now it's dependent on the parameter `training`.  Hence `predict(ntree_limit=0)` should has exact same result with `predict(ntree_limit=boosted_rounds)`.

**Comment by Kodiologist** [comment]
> You can get the old behaviour by setting `predict(training=True)`.

I don't seem to obtain this. On master, adding `training = T` to the `predict` call in the above code still gives me the new prediction vector, which is different from both of the prediction vectors on 7b65698.

**Comment by trivialfis** [comment]
@Kodiologist That might be a R issue, let me take a look.  I ran the following code to verify my comment above, could you please take a look and see if I'm misunderstanding some of your questions?

``` Python
'''Basic running script.'''

import numpy as np
import xgboost

np.random.seed(1994)

kRows = 64
kCols = 8

X = np.random.randn(kRows, kCols)
y = np.random.randn(kRows)

dtrain = xgboost.DMatrix(X, y)

bst = xgboost.train({
    'booster': 'dart',
    'verbosity': 1,
    'tree_method': 'exact',
    'rate_drop': '0.5'
},
                    dtrain,
                    num_boost_round=32)

prediction_0 = bst.predict(dtrain)
print(prediction_0)
prediction_1 = bst.predict(dtrain, ntree_limit=0)
print(prediction_1)

assert np.all(prediction_0 == prediction_1)

prediction_3 = bst.predict(dtrain, training=True)
print(prediction_3)

assert np.all(prediction_0 != prediction_3)
```

**Comment by trivialfis** [comment]
Good to know that I can do this:

``` R
library(xgboost, lib.loc = "/home/hippo/Workspace/xgboost/mylib")
```

Thanks!  That's gonna help a lot.

**Comment by trivialfis** [comment]
@Kodiologist Okay, I found the cause.  My mistake.   Will fix it.

**Comment by Kodiologist** [comment]
Fantastic, thank you.
